### PR TITLE
Added FranchiseFactory balances invariant test

### PR DESCRIPTION
**Pull request by jferas**
Added invariant that token totalSupply must match the total of all franchisers and recalled franchiser balances

### CODE REVIEW

File: test/handlers/FranchiserFactoryHandler.sol [hunk 1]
```diff
@@ -15,18 +15,24 @@ contract FranchiserFactoryHandler is Test {
     // Handler ghost array to contain all the funded franchisers created by handler_fund
     Franchiser[] public fundedFranchisers;
 
+    // Handler ghost array to contain all of the delegators that have recalled their franchisers
+    address[] public recalledDelegators;
```
Comment (wildmolasses) [inline test/handlers/FranchiserFactoryHandler.sol]: suggest making this an [AddressSet](https://github.com/horsefacts/weth-invariant-testing/blob/main/test/helpers/AddressSet.sol) -- will remove LOC like `isRecalledDelegatorInArray`

Comment (jferas) [reply test/handlers/FranchiserFactoryHandler.sol]: Thanks @wildmolasses the use of and AddressSet was also helpful for tracking the created Franchisers. 8ac4fb15dcd789ba189b7832b14814415b7ada53

Comment (jferas) [reply test/handlers/FranchiserFactoryHandler.sol]: Going to create a handler for fundMany.. which will simply create more Franchisers at once, and still keep the 2 invariants valid.

File: test/FranchiserFactory.invariants.t.sol [hunk 2]
```diff
@@ -16,14 +16,31 @@ contract FranchiseFactoryInvariantTest is Test {
...
+        selectors[5] = FranchiserFactoryHandler.handler_permitAndFundMany.selector;
         targetSelector(FuzzSelector({addr: address(handler), selectors: selectors}));
         targetContract(address(handler));
     }
 
     function invariant_Franchiser_code_size_non_zero() external view {
         assertGt(address(handler.franchiser()).code.length, 0);
     }
+
+    function invariant_FranchiserAndDeployersRecalledBalanceSumMatchesTotalSupply() external view {
```
Comment (alexkeating) [inline test/FranchiserFactory.invariants.t.sol]:
```suggestion
    function invariant_Franchiser_and_deployers_recalled_balance_sum_matches_total_supply() external view {
```

Comment (alexkeating) [reply test/FranchiserFactory.invariants.t.sol]: What is the "deployer"?

File: test/handlers/FranchiserFactoryHandler.sol [hunk 3]
```diff
@@ -2,63 +2,197 @@
...
 import {Franchiser} from "src/Franchiser.sol";
 
 contract FranchiserFactoryHandler is Test {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
     FranchiserFactory public factory;
     Franchiser public franchiser;
+    VotingTokenConcrete public votingToken;
+
+    mapping(bytes32 => uint256) public calls;
```
Comment (alexkeating) [inline test/handlers/FranchiserFactoryHandler.sol]: Named params would be clearer here

File: test/handlers/FranchiserFactoryHandler.sol [hunk 4]
```diff
@@ -2,63 +2,197 @@
...
-        fundedFranchisers.push(franchiser);
+        fundedFranchisers.add(address(franchiser));
+    }
+
+    function handler_fundMany(address _delegator, address[] memory _delegatees, uint256 _baseAmount)
+        external
+        countCall("handler_fundMany")
+    {
+        uint256 _numberOfDelegatees = _delegatees.length;
+        _baseAmount = _bound(_baseAmount, 1, 10_000e18);
```
Comment (alexkeating) [inline test/handlers/FranchiserFactoryHandler.sol]: Maybe bound this to an upper bound of 10 million

File: test/handlers/FranchiserFactoryHandler.sol [hunk 5]
```diff
@@ -2,63 +2,197 @@
...
+        votingToken.mint(_delegator, _totalAmountToMintAndApprove);
+        vm.startPrank(_delegator);
+        votingToken.approve(address(factory), _totalAmountToMintAndApprove);
+
+        // clear the storage of the lastFundedFranchisersArray and create a new one with call to fundMany
+        delete lastFundedFranchisersArray;
+        lastFundedFranchisersArray = factory.permitAndFundMany(_delegatees, _amountsForFundMany, _deadline, _v, _r, _s);
+        vm.stopPrank();
+
+        // add the created franchisers to the fundedFranchisers AddressSet for tracking totals invariants
```
Comment (alexkeating) [inline test/handlers/FranchiserFactoryHandler.sol]:
```suggestion
        // add the created franchisers to the fundedFranchisers AddressSet for tracking total invariants
```

File: test/handlers/FranchiserFactoryHandler.sol [hunk 6]
```diff
@@ -2,63 +2,197 @@
...
+            return;
+        }
+        _fundedFranchiserIndex = bound(_fundedFranchiserIndex, 0, fundedFranchisers.length() - 1);
+        Franchiser _fundedFranchiser = Franchiser(fundedFranchisers.at(_fundedFranchiserIndex));
         address _delegatee = _fundedFranchiser.delegatee();
         address _delegator = _fundedFranchiser.delegator();
-        vm.prank(_delegator);
+        uint256 _amount = votingToken.balanceOf(address(_fundedFranchiser));
+
+        // do the recall of delegated funds then move the recalled funds to the targetAddressForRecalledFunds
```
Comment (alexkeating) [inline test/handlers/FranchiserFactoryHandler.sol]:
```suggestion
        // recall of delegated funds then move the recalled funds to the targetAddressForRecalledFunds
```

Comment (jferas) [reply test/handlers/FranchiserFactoryHandler.sol]: Done: d22055bb4ad63bc533679b3cae384b41e0c89bc1

File: test/handlers/FranchiserFactoryHandler.sol (hunk 5 above)
Comment (jferas) [reply test/handlers/FranchiserFactoryHandler.sol]: Done: d22055bb4ad63bc533679b3cae384b41e0c89bc1

File: test/handlers/FranchiserFactoryHandler.sol (hunk 4 above)
Comment (jferas) [reply test/handlers/FranchiserFactoryHandler.sol]: Done: 9211e4a5554f20c1e9f74eb5a031d4a5a817f049

File: test/handlers/FranchiserFactoryHandler.sol (hunk 3 above)
Comment (jferas) [reply test/handlers/FranchiserFactoryHandler.sol]: Done: 5ab7405ae8e5300d5463faa519a6756c3d95c563

File: test/FranchiserFactory.invariants.t.sol (hunk 2 above)
Comment (jferas) [reply test/FranchiserFactory.invariants.t.sol]: Done, and took deployer out of name: f6d0fd2816c6d3b5c46dc2b0a90b1da4b0559356

**Review by alexkeating — APPROVED** [review]

**Pull request merged**